miniconf/src/packed.rs

```diff
@@ -0,0 +1,219 @@
+    }
+
+    /// Clear and discard all bits stored.
+    #[inline]
+    pub fn clear(&mut self) {
+        *self = Self::EMPTY;
+    }
+
+    /// Number of bits stored.
+    #[inline]
+    pub const fn len(&self) -> u32 {
+        Self::CAPACITY - self.0.trailing_zeros()
+    }
+
+    /// Return the representation aligned to the LSB with the marker bit
+    /// moved from the LSB to the MSB.
+    #[inline]
+    pub fn into_lsb(&self) -> NonZeroUsize {
+        // Note(unwrap): we ensure there is at least the marker bit set
+        NonZeroUsize::new(((self.get() >> 1) | (1 << Self::CAPACITY)) >> self.trailing_zeros())
+            .unwrap()
+    }
+
+    /// Build a `Packed` from a LSB-aligned representation with the marker bit
+    /// moved from the MSB the LSB.
+    #[inline]
+    pub fn from_lsb(value: NonZeroUsize) -> Self {
+        // Note(unwrap): we ensure there is at least the marker bit set
+        Self::new(((value.get() << 1) | 1) << value.leading_zeros()).unwrap()
+    }
+
+    /// Return the number of bits required to represent `num`.
+    ///
+    /// Ensures that at least one bit is allocated.
+    #[inline]
+    pub fn bits_for(num: usize) -> u32 {
+        (Self::BITS - num.leading_zeros()).max(1)
+    }
+
+    /// Remove the given number of MSBs and return them.
+    ///
+    /// If the value does not contain sufficient bits
+    /// it is left unchanged and `None` is returned.
+    ///
+    /// # Args
+    /// * `bits`: Number of bits to pop. `bits <= Self::CAPACITY`
+    #[inline]
+    pub fn pop_msb(&mut self, bits: u32) -> Option<usize> {
+        let s = self.get();
+        if let Some(new) = Self::new(s << bits) {
+            self.0 = new.0;
+            Some((s >> (Self::CAPACITY - bits)) >> 1)
+        } else {
+            None
+        }
+    }
+
+    /// Push the given number `bits` of `value` as new LSBs.
+    ///
+    /// Returns the remaining number of unused bits on success.
+    ///
+    /// # Args
+    /// * `bits`: Number of bits to push. `bits <= Self::CAPACITY`
+    /// * `value`: Value to push. `value >> bits == 0`
+    #[inline]
+    pub fn push_lsb(&mut self, bits: u32, value: usize) -> Option<u32> {
+        debug_assert_eq!(value >> bits, 0);
+        let mut n = self.trailing_zeros();
+        let m = 1 << n;
+        if let Some(marker) = Self::new(m >> bits) {
+            n -= bits;
+            self.0 = (self.get() ^ m) | ((value << n) << 1) | marker.0;
+            Some(n)
+        } else {
+            None
+        }
+    }
+}
+
+impl Keys for Packed {
+    type Item = usize;
+
+    #[inline]
+    fn next(&mut self, len: usize) -> Option<Self::Item> {
+        self.pop_msb(Self::bits_for(len.saturating_sub(1)))
+    }
+}
+
+impl IntoKeys for Packed {
+    type IntoKeys = Self;
+
+    #[inline]
+    fn into_keys(self) -> Self::IntoKeys {
+        self
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test() {
+        let t = [1usize, 3, 4, 0, 1];
```

Comment:
It would be good to have some magic number pre-calculated value tests as well to show that it works with only one side of the API. I.e. provide a packed bitstring with known properties and then pop the known bits out of it.

Reply:
We could have some weird symmetric bug (i.e. happens on both input and output) that we aren't finding because of this.

Reply:
Done in the struct doctest example.